api/src/main/java/jakarta/data/metamodel/StaticMetamodel.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023,2024 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.metamodel;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import jakarta.data.Sort;
+
+/**
+ * <p>Annotates a class to serve as a static metamodel for an entity,
+ * enabling type-safe access to entity attribute names and related objects,
+ * such as {@link Sort}s for an attribute.</p>
+ *
+ * <p>For example, for the following entity,</p>
+ *
+ * <pre>
+ * &#64;Entity
+ * public class Person {
+ *     &#64;Id
+ *     public long ssn;
+ *
+ *     &#64;Embedded
+ *     public Name name;
+ *
+ *     public int yearOfBirth;
+ * }
+ *
+ * &#64;Embeddable
+ * public class Name {
+ *     public String first;
+ *     public String last;
+ * }
+ * </pre>
+ *
+ * <p>You can define a static metamodel as follows,</p>
+ *
+ * <pre>
+ * &#64;StaticMetamodel(Person.class)
+ * public class _Person {
+ *     // These can be uninitialized and non-final if you don't need to access them from annotations.
+ *     public static final String SSN = "ssn";
+ *     public static final String NAME = "name";
+ *     public static final String NAME_FIRST = "name.first";
```

Comment:
What happens with Entity with relationship?

For example,

```java
class Person {

private Address address;
}

class Address {
private String street;
}
```
Thus, the metadata:

```java

class PersonMetadata {
public static final String Address_STREET = "address.street";
}

class AddressMetadata {
public static final String STREET = "name";
}
```

Please ignore the terminology. My point is. This constant might have a substantial recursive constant.

I would skip this constant.

For example, Jakarta Persistence does not have this:

https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#staticmetamodel-annotation

Reply:
> What happens with Entity with relationship?

An entity only has these fields for its _direct_ members. So no `HEADQUARTER_STREET`, etc.

[Note that I guess we didn't say anything about embeddables yet, but in JPA embeddables have their own static metamodel types.]

> Jakarta Persistence does not have this

It's coming in JPA 3.2

Reply:
Your example mostly looks good but has an error/typo here:

> ```java
> class AddressMetadata {
> public static final String STREET = "name";
> }
> ```

The value should be "street".

> Please ignore the terminology. My point is. This constant might have a substantial recursive constant.
> As much as we have a relationship, it might grow. E.g.: `person.address.country.city.street`
>
> I would skip this constant.

Your particular hierarchy doesn't make any sense. You wouldn't find a city embeddable inside of the country or a street field on a city.  Let's use `person.address.city.state.country` as a better example.
If the user defines their entities and embeddable objects that way, it certainly could have a lengthy relationship. But that isn't a problem of the static metamodel. The Jakarta Data specification already defines this behavior with embeddables.  A user can already do this if they want to:

```
people.findByYearOfBirth(int birthYear, Sort.asc("person.address.city.state.country"), ...);
```

The only part the static metamodel is adding is allowing them to do it in a type safe way.

The example with that many levels of embeddables isn't as realistic. However, one or two levels of embeddables is something users will want to do.  This is something we support in the spec, and I don't see why we should skip over having a type safe solution for it like we do for other usage.

> For example, Jakarta Persistence does not have this

That's fine. We don't need to match Persistence for the sake of copying. We should include it because referencing embeddables in this way is part of the Jakarta Data programming model and we should make our whole programming model type safe.

Reply:
For the record I'm not sure this is necessary:

    public static final String NAME_FIRST = "name.first";

But I need to think it through some more.

Reply:
> The example with that many levels of embeddables isn't as realistic.

Unfortunately, it is more natural than we wish. But let me create a real scenario here, and I will simplify it.

Unfortunately, it happens more than I wish, for example, in e-commerce:

```java
class Customer {
    private long id;
    private Address billingAddress;
    private List<Order> orders;
}

class Order {
    private long id;
    private Customer customer;
    private List<Product> products;
    private Address shipping;
    private Payment payment;

}

class Address {

}

class Payment {
Address address;
}
```


So, there are two different paths for Addresses, where to client this is for billing and the order the address is to shipping the product:
* `billingAddress.street`
* `shipping.street`
My point is only the static and how to solve this point.


⚠️ We have an even worse scenario in finance, especially with credit-card providers integration.

Reply:
> So, there are two different paths for Addresses, where to client this is for billing and the order the address is to shipping the product:
> 
>     * `billingAddress.street`
> 
>     * `shipping.street`
>       My point is only the static and how to solve this point.

Otavio, yes, that's an excellent scenario. But why are you trying to use that scenario as an argument against the spec updates in this pull?  This pull provides the type-safe solution to your scenario.  If you follow what is in this pull and apply it to your example scenario, everything is very straightforward,

```
@StaticMetamodel(Customer.class)
class _Customer {
    public static final String ID = "id";
    public static final String BILLINGADDRESS = "billingAddress";
    public static final String BILLINGADDRESS_STREET = "billingAddress.street";
    public static final String ORDERS = "orders";
}

@StaticMetamodel(Order.class)
class _Order {
    public static final String ID = "id";
    public static final String CUSTOMER = "customer";
    public static final String CUSTOMER_BILLINGADDRESS = "customer.billingAddress";
    public static final String CUSTOMER_BILLINGADDRESS_STREET = "customer.billingAddress.street";
    public static final String PRODUCTS = "products";
    public static final String SHIPPING = "shipping";
    public static final String SHIPPING_STREET = "shipping.street";
    public static final String PAYMENT = "payment";
    public static final String PAYMENT_ADDRESS = "payment.address";
    public static final String PAYMENT_ADDRESS_STREET = "payment.address.street";
}

@StaticMetamodel(Address.class)
class _Address {
    public static final String STREET = "street";
}

@StaticMetamodel(Payment.class)
class _Payment {
    public static final String ADDRESS = "address";
    public static final String ADDRESS_STREET = "address.street";
}
```

Example usage:

```
orders.findAll(Sort.asc(_Order.CUSTOMER_BILLINGADDRESS_STREET));
orders.findAll(Sort.asc(_Order.SHIPPING_STREET));
orders.findAll(Sort.asc(_Order.PAYMENT_ADDRESS_STREET));
```

The above usage equates to what you could already write as strings without the static metamodel,

```
orders.findAll(Sort.asc("customer.billingAddress.street"));
orders.findAll(Sort.asc("shipping.street"));
orders.findAll(Sort.asc("payment.address.street"));
```

If the relationship was embeddables, this will all work cleanly. If it is some other sort of relationship that isn't required by Jakarta Data, then it's up to the provider what will happen.  This pull doesn't change any of that.  The only part added in this pull is the ability to use type safe values in place of the strings.

That said, if for some reason everyone else is against specifying and standardizing how to define the type constants and attribute fields for embeddable attributes, we could certainly leave that part of the static metamodel for now.

Reply:
OK, so now I see the thing I missed.

In Hibernate, the equivalent type to `Sort` is parameterized by the entity type. So it's `Sort<Customer>` which allows a query method to accept only the right kind of `Sort` objects.

We don't have that here yet. since `Sort` is not declared as a parameterized type.

So what I would like us to work towards is:

```java
repository.allCustomers(_Customer.name.asc())
```

which we already have, but I would like it to be more typesafe, and:

```java
repository.allCustomers(_Address.name.of(_Customer.billingAddress).asc())
```

or something like that ... just trying things out here.

@njr-11 @otaviojava Do you follow what I'm trying to get at here?

Reply:
i.e you would declare things like:

```java
Employee[] findByYearHired(int yearYired, Limit maxResults, Sort<Employee>... sortBy);
```

Reply:
> i.e you would declare things like:
> 
> ```java
> Employee[] findByYearHired(int yearYired, Limit maxResults, Sort<Employee>... sortBy);
> ```
This is nice because during development time it helps prevent the error of accidentally using a Sort that was intended for a different entity type.

> ```java
> repository.allCustomers(_Address.name.of(_Customer.billingAddress).asc())
> ```

That's a good example of where a parameterized sort would be necessary to disambiguate which entity the query pertains to, which is another scenario where the parameterized sort would be useful.  Let's open a separate issue for this idea because it applies regardless of whether we have the static metamodel.

I should have written up my earlier example with both the String constants and the Attribute fields because the latter should already cover the type-safe sort objects:

```
repository.allCustomers(_Customer.billingAddress_street.asc())
```

Relying on the following metamodel classes...

```
@StaticMetamodel(Address.class)
class _Address {
    public static final String STREET = "street";

    public static volatile SortableAttribute street;
}

@StaticMetamodel(Customer.class)
class _Customer {
    public static final String ID = "id";
    public static final String BILLINGADDRESS = "billingAddress";
    public static final String BILLINGADDRESS_STREET = "billingAddress.street";

    public static volatile SortableAttribute id;
    public static volatile Attribute billingAddress;
    public static volatile SortableAttribute billingAddress_street;
}
```

It should be noted that if we add parameterized sort, then SortableAttribute would likely be parameterized as well,
```
    public static volatile SortableAttribute<Customer> id;
```
so that _Customer.id.asc() will return a Sort<Customer> that forces type-safe usage.

Reply:
I am fine to merge the PR, but we need to cover all scenarios once we include this metadata, furthermore, it should cover both SQL and NoSQL databases. Including the feature is the first step; make sure it works is the second step.

Reply:
About the path, I would go to the same path that we do in code as metadata, thus:

`_Customer. _Address.street.asc()`

Reply:
Yeah, I guess that's about right @otaviojava .... except probably this: `_Customer.address.street.asc()`

Reply:
Excellent - I think that's exactly how we defined it, so we just need to confirm that what we wrote is sufficiently clear.

Reply:
Disregard my comment above about that exactly matching how we defined it.  I misread it as `_Customer.address_street.asc()` (how we defined it), but the difference is in the `.` vs `_` between `address` and `street`.  It does look nicer the way you wrote it.  However, for that to work, the `_Customer.address` Attribute would also need to also be a static metamodel class, so it would need to have a type that is both `Attribute` and `_Address`.

```java
@StaticMetamodel(Customer.class)
public class _Customer {
    public static final String ID = "id";
    public static final String BILLINGADDRESS = "billingAddress";
    public static final String BILLINGADDRESS_STREET = "billingAddress.street";

    public static volatile SortableAttribute id;
    public static volatile _Address billingAddress; // <-- Both Attribute and _Address
}

@StaticMetamodel(Address.class)
public class _Address implements Attribute { // <-- added implements
    public static final String STREET = "street";

    public static volatile SortableAttribute street;

    @Override
    public String name() { // <-- who provides/implements the Attribute methods?
}
```

This approach seems to make things more complex and has some issues that would need to be worked through. It might be possible for the metamodel class for the embeddable to become abstract and then avoid declaring the Attribute methods, but ultimately the method does need to be implemented, and wherever that happens will run into the problem that the name() is different depending on which entity embeds it, because they can each name the Address field whatever they like.